cmd/common/oci.go

```diff
@@ -0,0 +1,153 @@
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/logger"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+	clioperator "github.com/inspektor-gadget/inspektor-gadget/pkg/operators/cli"
+	ocioperator "github.com/inspektor-gadget/inspektor-gadget/pkg/operators/oci"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/runtime"
+)
+
+func NewRunCommand(rootCmd *cobra.Command, runtime runtime.Runtime, hiddenColumnTags []string) *cobra.Command {
+	runtimeGlobalParams := runtime.GlobalParamDescs().ToParams()
+	runtime.Init(runtimeGlobalParams)
+
+	ociParams := ocioperator.OciHandler.ParamDescs().ToParams()
+
+	// Add operator global flags
+	operatorsGlobalParamsCollection := operators.GlobalParamsCollection()
+
+	cmd := &cobra.Command{
+		Use:          "run-experimental",
+		Short:        "run gadget",
+		SilenceUsage: true, // do not print usage when there is an error
+		// We have to disable flag parsing in here to be able to handle certain
+		// flags more dynamically and have `--help` also react to those changes.
+		// Instead, we need to manually
+		// * call cmd.ParseFlags()
+		// * handle `--help` after changing the params dynamically
+		// * handle everything that could have been handled inside
+		//   `PersistentPreRun(E)` of a parent cmd, as the flags wouldn't have
+		//   been parsed there (e.g. --verbose)
+		DisableFlagParsing: true,
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			// Manually register the clioperator
+			clioperator.Register()
+
+			err := runtime.Init(runtimeGlobalParams)
+			if err != nil {
+				return fmt.Errorf("initializing runtime: %w", err)
+			}
+			defer runtime.Close()
+
+			// we need to re-enable flag parsing, as utils.ParseEarlyFlags() would
+			// not do anything otherwise
+			cmd.DisableFlagParsing = false
+
+			// Parse flags that are known at this time, like the ones we get from the gadget descriptor
+			if err := utils.ParseEarlyFlags(cmd, args); err != nil {
+				return err
+			}
+
+			// gadget parameters that are only available after contacting the server
+			gadgetParams := make(params.Params, 0)
+
+			// Before running the gadget, we need to get the gadget info to perform
+			// different tasks like creating the parser and setting flags for the
+			// gadget's parameters.
+			actualArgs := cmd.Flags().Args()
+
+			gadgetCtx := gadgetcontext.NewOCI(
+				context.Background(),
+				actualArgs[0],
+				logger.DefaultLogger(),
+				params.Collection{
+					"oci": ociParams,
+				},
+			)
+
+			// Fetch gadget information; TODO: this can potentially be cached
+			info, err := runtime.GetOCIGadgetInfo(gadgetCtx)
+			if err != nil {
+				return fmt.Errorf("fetching gadget information: %w", err)
+			}
+
+			for _, p := range info.Params {
+				gadgetParams.Add(api.ParamToParamDesc(p).ToParam())
+			}
+
+			AddFlags(cmd, &gadgetParams, nil, runtime)
```

Comment:
Wow, that diagram is super detailed, I love it!

I'm sure we can (and should) optimize a couple of things to be postponed until after the gadget is actually started (maps, tailcalls, etc. - I'll look into it - this should also take care of the leaks, then), however, the initialization with Params / DataSources is actually needed:

Params can influence which DataSources are created, DataSources can influence other DataSources (and fields of course) - DataSources can potentially influence which Params are created. So they're heavily interweaved.

If any of the heavier eBPF initialization affects DataSources/Params, it also has to be done in the initialization phase - if not, it simply doesn't belong there (at least not in the final version).

Metadata passing isn't needed right now, but is kept as opaque info that can (also) be consumed by an external client. Maybe we should only deliver that on-demand, will look into it as well (though maybe after this PR).

All of this can and should be cached - but only on the server side. The idea was to do it in a follow-up PR.

In regard to the client side: operators can (and should) also run locally - everything that just transforms existing fields for example should be done on the client to avoid sending more data than necessary. Thus, they also need information about the DataSource. (Note: write access to DataSources on local operators has not yet been implemented; this requires additional safeguards on the DataSources)

Reply:
So I looked into this a bit more: I think all the heavy lifting can go into `Start()` easily. I already changed the order of initialization (DataOperators first, then LayerOperators - haven't pushed yet, though), so that should work well.

Reply:
> Wow, that diagram is super detailed, I love it!
>
🙂 ... But navigating through it in GH editor is not the best, I prefer [this live editor](https://mermaid.live/).

> Params can influence which DataSources are created, DataSources can influence other DataSources (and fields of course) - DataSources can potentially influence which Params are created. So they're heavily interweaved.
>
Makes sense. Let's keep DataSources registration in the `Instantiate` then.

> Metadata passing isn't needed right now, but is kept as opaque info that can (also) be consumed by an external client.
>
I don't see a use case for this, do you?

> Maybe we should only deliver that on-demand, will look into it as well (though maybe after this PR).
>
Sure. The only thing I think we should manage in this PR about this comment is the leaking.

> All of this can and should be cached - but only on the server side. The idea was to do it in a follow-up PR.
>
Given that DataSources need to be registered in the initial phase, then yes. I agree. We need to cache it on the server side. But, why not also on the client side to avoid sending the `EventTypeGadgetInfo` message with the same info (mainly DataSources)? DataSource won't change from `GetOCIGadgetInfo` to `RunOCIGadget`, so why not cache it too on the client side?

> So I looked into this a bit more: I think all the heavy lifting can go into Start() easily. I already changed the order of initialization (DataOperators first, then LayerOperators - haven't pushed yet, though), so that should work well.
>
I had thought that `runtime.local.GetOCIGadgetInfo` could `Instantiate` but not `Prepare` operators. Postponing the `Prepare` should also postpone the expensive operations. Did I miss something? Just to be sure I understood the flow, not because I think it's a better solution 😝.

Reply:
> > Wow, that diagram is super detailed, I love it!
> 
> 🙂 ... But navigating through it in GH editor is not the best, I prefer [this live editor](https://mermaid.live/).

I used that as well for mine 😄 

> > Params can influence which DataSources are created, DataSources can influence other DataSources (and fields of course) - DataSources can potentially influence which Params are created. So they're heavily interweaved.
> 
> Makes sense. Let's keep DataSources registration in the `Instantiate` then.
> 
> > Metadata passing isn't needed right now, but is kept as opaque info that can (also) be consumed by an external client.
> 
> I don't see a use case for this, do you?

Mainly for third party consumers - I don't have a use case now, but I'm sure people will come up with many :) And as it's cheap to do, I'd like to support it from the start. (It's basically again the question: how can I store something inside a gadget metadata file that I can easily access from the client -> this is the solution; annotations might not be enough / too tied to our structures)

> 
> > Maybe we should only deliver that on-demand, will look into it as well (though maybe after this PR).
> 
> Sure. The only thing I think we should manage in this PR about this comment is the leaking.

👍 
 
> > All of this can and should be cached - but only on the server side. The idea was to do it in a follow-up PR.
> 
> Given that DataSources need to be registered in the initial phase, then yes. I agree. We need to cache it on the server side. But, why not also on the client side to avoid sending the `EventTypeGadgetInfo` message with the same info (mainly DataSources)? DataSource won't change from `GetOCIGadgetInfo` to `RunOCIGadget`, so why not cache it too on the client side?

Mainly because `GetOCIGadgetInfo` is requested from _one_ server, while `RunOCIGadget` uses all targets. I've written some experimental code on synchronizing/validating the `EventTypeGadgetInfo` against each other, but haven't pushed it, yet. Will do so after a bit more testing.

> > So I looked into this a bit more: I think all the heavy lifting can go into Start() easily. I already changed the order of initialization (DataOperators first, then LayerOperators - haven't pushed yet, though), so that should work well.
> 
> I had thought that `runtime.local.GetOCIGadgetInfo` could `Instantiate` but not `Prepare` operators. Postponing the `Prepare` should also postpone the expensive operations. Did I miss something? Just to be sure I understood the flow, not because I think it's a better solution 😝.

I'll re-check whether `Prepare` is still necessary (it might make sense to keep a couple of lifecycle functions to hook into things at different phases); the main reason for now was to keep it similar to how operators work right now: `Instantiate` should be as cheap as possible, `Prepare` can do the rest that is necessary for initialization.

Reply:
> We are passing the same URL to all targets, do we really need to verify all of them download the same gadget?

Out-of-sync could happen for quite a couple of reasons, here are some of them:
* different gadget versions (e.g. using 'latest' but one node has been added after a new version of the gadget was released)
* different IG versions (could be prevented)
* different operators activated

So yeah, I think it's necessary to always send the GadgetInfo.